reconcile/status_board.py

```diff
@@ -0,0 +1,216 @@
+
+        for p in products:
+            if not p.id:
+                logging.error(f'Product "{p.name}" has no id')
+                continue
+            p.applications = [
+                Application(**a) for a in get_product_applications(ocm_api, p.id)
+            ]
+
+        return products
+
+    @staticmethod
+    def get_diff(
+        desired_product_apps: dict[str, set[str]],
+        current_products_applications: list[Product],
+    ) -> list[StatusBoardHandler]:
+        return_list: list[StatusBoardHandler] = []
+        current_products = [p.name for p in current_products_applications]
+        for product_name, app_names in desired_product_apps.items():
+            if product_name not in current_products:
+                return_list.append(
+                    StatusBoardHandler(
+                        action="create",
+                        status_board_object=Product(
+                            name=product_name, fullname=product_name, applications=[]
+                        ),
+                    )
+                )
+            else:
+                # else, only create App if Product exists
+                product_object = next(
+                    p for p in current_products_applications if p.name == product_name
+                )
+                current_applications = [
+                    a.name for a in product_object.applications or []
+                ]
+                for app_name in app_names:
+                    if app_name not in current_applications:
+                        return_list.append(
+                            StatusBoardHandler(
+                                action="create",
+                                status_board_object=Application(
+                                    name=app_name,
+                                    fullname=f"{product_name}/{app_name}",
+                                    product=product_object,
+                                ),
+                            )
+                        )
+
+        for product in current_products_applications:
+            if product.name not in desired_product_apps:
+                # if product was removed, we have to delete all apps
+                for application in product.applications or []:
+                    return_list.append(
+                        StatusBoardHandler(
+                            action="delete", status_board_object=application
+                        )
+                    )
+                return_list.append(
+                    StatusBoardHandler(action="delete", status_board_object=product)
+                )
+            else:
+                for app in product.applications or []:
+                    if app.name not in desired_product_apps.get(product.name, []):
+                        return_list.append(
+                            StatusBoardHandler(action="delete", status_board_object=app)
+                        )
```

Comment:
I'm selling new [generic differ](https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/differ.py).

```python
        current_products = {p.name: p for p in current_products_applications}
        diff_result = diff_mappings(
            current_products,
            desired_product_apps,
            equal=lambda c, d: {a.name for a in c.applications} == d,
        )

        for product_name in diff_result.add:
            return_list.append(
                StatusBoardHandler(
                    action="create",
                    status_board_object=Product(
                        name=product_name, fullname=product_name, applications=[]
                    ),
                )
            )

        for diff_pair in diff_result.change.values():
            desired_app_names = diff_pair.desired
            current_product = diff_pair.current
            current_app_names = {a.name for a in current_product.applications or []}
            for app_name in desired_app_names - current_app_names:
                return_list.append(
                    StatusBoardHandler(
                        action="create",
                        status_board_object=Application(
                            name=app_name,
                            fullname=f"{current_product.name}/{app_name}",
                            product=current_product,
                        ),
                    )
                )
            for app in current_product.applications or []:
                if app.name not in desired_app_names:
                    return_list.append(
                        StatusBoardHandler(action="delete", status_board_object=app)
                    )

        for product in diff_result.delete.values():
            for application in product.applications or []:
                return_list.append(
                    StatusBoardHandler(action="delete", status_board_object=application)
                )
            return_list.append(
                StatusBoardHandler(action="delete", status_board_object=product)
            )

```

Reply:
Awesome stuff, looks like I can drop it in without issues. I'll give it a shot